contribs/gmf/src/directives/print.js

```diff
@@ -0,0 +1,803 @@
+
+  var legend = this.isAttributeInCurrentLayout_('legend');
+  this.fields.legend = legend !== null ?
+      this.fields.legend || true : undefined;
+
+  this.fields.scales = clientInfo['scales'] || [];
+  this.fields.dpis = clientInfo['dpiSuggestions'] || [];
+
+  var mapSize = this.map.getSize();
+  var viewResolution = this.map.getView().getResolution();
+  this.fields.scale = this.getOptimalScale_(mapSize, viewResolution);
+
+  this.fields.dpi =
+      (this.fields.dpi && this.fields.dpis.indexOf(this.fields.dpi) > 0) ?
+      this.fields.dpi : this.fields.dpis[0];
+
+  this.fields.formats = {};
+  this.formats_.forEach(function(format) {
+    this.fields.formats[format] = true;
+  }.bind(this));
+};
+
+
+/**
+ * Update customs fields with gmfx.Customfield to be able to generate a form
+ * from a custom GMF print v3 configuration.
+ * @private
+ */
+gmf.PrintController.prototype.updateCustomFields_ = function() {
+  var name, rawType, value, type;
+  if (!this.fields.customs) {
+    this.fields.customs = [];
+  }
+  var customs = this.fields.customs;
+  var previousCustoms = customs.splice(0, customs.length);
+
+  // The attributes without 'clientParams' are the custom fields (user-defined).
+  this.layout_.attributes.forEach(function(attribute) {
+    if (!attribute['clientParams']) {
+      name = '' + attribute.name;
+      value = attribute.default;
+
+      // Try to use existing form field type
+      rawType = '' + attribute.type;
+      switch (rawType) {
+        case 'String':
+          type = (name === 'comments') ? 'textarea' : 'text';
+          break;
+        case 'Boolean':
+          type = 'checkbox';
+          break;
+        case 'Number':
+          type = 'number';
+          break;
+        default:
+          type = rawType;
+      }
+
+      // If it exists use the value of previous same field.
+      previousCustoms.forEach(function(c) {
+        if (c.name === name && c.type === type) {
+          return value = c.value;
+        }
+      });
+
+      this.fields.customs.push(/** gmfx.CustomField */ ({
+        name: name,
+        type: type,
+        value: value
+      }));
+    }
+  }.bind(this));
+};
+
+
+/**
+ * Return a capabilities 'attribute' object corresponding to the given name.
+ * @param {string} name Name of the attribute to get.
+ * @return {Object|null} corresponding attribute or null.
+ * @private
+ */
+gmf.PrintController.prototype.isAttributeInCurrentLayout_ = function(name) {
+  var attr = null;
+  this.layout_.attributes.forEach(function(attribute) {
+    if (attribute.name === name) {
+      return attr = attribute;
+    }
+  });
+  return attr;
+};
+
+
+/**
+ * Getter setter to update or get the current rotation value. Param and result
+ *     are in degree.
+ * @param {number=} opt_rotation The optional new rotation value.
+ * @return {number} The new value of rotation;
+ * @export
+ */
+gmf.PrintController.prototype.getSetRotation = function(opt_rotation) {
```

Comment:
Please search "gettersetter" on this page: https://docs.angularjs.org/api/ng/directive/ngModel